pydrive2/fs/spec.py

```diff
@@ -534,13 +534,22 @@ def _gdrive_upload_fobj(self, title, parent_id, stream, callback=None):
         return item
 
     def cp_file(self, lpath, rpath, **kwargs):
-        """In-memory streamed copy"""
-        with self.open(lpath) as stream:
-            # IterStream objects doesn't support full-length
-            # seek() calls, so we have to wrap the data with
-            # an external buffer.
-            buffer = io.BytesIO(stream.read())
-            self.upload_fobj(buffer, rpath)
+        """copy a single file or folder"""
+
+        if self.exists(rpath):
+            raise FileExistsError(rpath)
+
+        file_id = self._get_item_id(lpath)
+
+        file = self.client.CreateFile({"id": file_id})
+
+        if file["mimeType"] == FOLDER_MIME_TYPE:
+            dst_id = self._get_item_id(self._parent(rpath))
+            self._gdrive_create_dir(dst_id, rpath.split("/")[-1])
+        else:
+            dst_parent_id = self._get_item_id(self._parent(rpath), create=True)
```

Comment:
what if `rpath` is a folder? what is the semantics? I would expect for this operation to copy file there. It looks it's not how it was done, but can we check `fsspec` docs and see what is expected?

Reply:
`create=True` - also need to check semantics of `fsspec` . I'm not sure if we should expect parents to be created.

Reply:
of the docs of fsspec is not specified how it should work in terms of paths: [docs](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.copy)

The way it's implemented now the name of the target need to be always specified, so `rpath` is a folder only in case `lpath` is also a folder and `recursive=True`.
If you want to copy a file and keep the same name you need to specify it on both `lpath` and `rpath`:

```python
fs.copy('root/tmp/a.pdf', 'root/tmp/folder/a.pdf', recursive=False)
```
I did not implement this behavior but is implemented in  `AbstractFileSystem.copy`

Reply:
so, `cp_file` is not public, right? 

can we check S3 implementation for a reference of `cp_file` to make sure regarding the semantics. This logic with directories looks weird a bit ...

Reply:
i checked with `s3fs` but they use the `AsyncFileSystem` and it's a little weird.

they have implemented a `_cp_file`: https://github.com/fsspec/s3fs/blob/18009c4ebdd19d688c3b10cd1dc7147dfaaae351/s3fs/core.py#L1651

and this get called by `_copy` in `AsyncFileSystem`: https://github.com/fsspec/filesystem_spec/blob/3f8c8e07219622e68e428214300a548ab7a2f8b7/fsspec/asyn.py#L344-L353

but than i did not find the connection to the public methods `copy` or `cp_file`.

another example i found is on `adlfs`: Filesystem interface to Azure-Datalake Gen1 and Gen2 Storage

https://github.com/fsspec/adlfs/blob/7a74d64d5a5332a4ee664a82322a6e16d60d88f8/adlfs/spec.py#L1628-L1644

in this case they implemented `_cp_file` and than renamed `cp_file = _cp_file`.